crates/bevy_ui/src/layout/ui_surface.rs

```diff
@@ -0,0 +1,995 @@
+use std::fmt;
+
+use taffy::prelude::LayoutTree;
+use taffy::Taffy;
+use thiserror::Error;
+
+use bevy_ecs::entity::{Entity, EntityHashMap, EntityHashSet};
+use bevy_ecs::prelude::Resource;
+use bevy_math::UVec2;
+use bevy_utils::default;
+use bevy_utils::hashbrown::hash_map::Entry;
+use bevy_utils::tracing::warn;
+
+use crate::layout::convert;
+use crate::{LayoutContext, LayoutError, Style};
+
+#[derive(Debug, Error)]
+/// Error wrapper for most actions performed in `UiSurface`
+pub enum UiSurfaceError {
+    #[error("Invalid state: {0}")]
+    /// Used whenever a creating function encounters a situation where the internal maps are not returning expected results.
+    /// Otherwise, where appropriate `EntityNotFound` or `TaffyError` are likely used.
+    InvalidState(&'static str),
+    #[error("Entity not found in internal map: {0}")]
+    /// Used for when an entity is not found in a map.
+    /// But only when it's not explicitly expected to exist by the function creating the error.
+    /// Otherwise, `InvalidState` is used.
+    EntityNotFound(Entity),
+    #[error("Taffy error: {0}")]
+    /// Simple wrapper for `TaffyError`.
+    TaffyError(#[from] taffy::error::TaffyError),
+    #[error("Iteration error {0:?}")]
+    /// Used in loops that attempt to continue despite encountering an error.
+    IterationError(Vec<UiSurfaceError>),
+    #[error("Error processing {0:?}")]
+    /// Wrapper error usually used in conjunction with `IterationError`.
+    /// Boxed tuple associates `Entity` to the error that was encountered when processing.
+    ///
+    // (boxed to allow self referencing)
+    ProcessingEntityError(Box<(Entity, UiSurfaceError)>),
+    #[error("Unreachable")]
+    /// Code paths that are deemed unreachable but avoid `unreachable!` to allow caller to handle and proceed
+    Unreachable,
```

Comment:
This... doesn't really make sense to me. Unreachable code by definition has no flow control paths leading to it. How is a caller expected to handle this error? Why is avoiding `unreachable!()` even a goal? You can't get rid of all possible panics, `std` is riddled with panicking API...

The same applies to internal state errors. If an internal invariant doesn't hold that is a bug, not an exception, and I don't understand how application code is expected to handle this. Panicking with an appropriate message is the most reasonable course of action in these scenarios IMO.

Reply:
Well, you're not wrong, and I agree. I didn't mean *all* panics, just the ones created by unwrap and such. However, I'm clearly getting carried away by including `Unreachable` in its scope. Thank you for calling me out on that early.

My original intentions were to bubble everything up to the caller (in this case, the `ui_layout_system`), expecting it to return a `Result` as well. Then that could just let unhandled errors trigger error logs (via `ui_layout_system.map(error)`) instead of crashing. That way, it could *maybe* allow other UI entities to finish processing. But that [increases the complexity as seen here](https://github.com/StrikeForceZero/bevy/compare/dev/bevy_ui/ui_surface_update_mappings..dev/bevy_ui/remove_unwrap#diff-91216818db3c390603f5a93d8a47d11e1a627d0c29578843800ea88849230b0cL404-L411). 

You make another excellent point about "how can a caller be expected to handle `InvalidState`?" 

Context: Right now, the internal maps are exposed as `pub(super)`. This could result in someone adding code outside the structs associated functions that interact directly with the internal map or the specific sequence of calls initiated by the calling system that creates an `InvalidState`. For now, `InvalidState` is a marker for something intended to be unreachable but not guaranteed. If the maps can be made private and the exposed API made to be infallible, then I think replacing `InvalidState` back to `panic!` or `unreachable!` depending on the context would be appropriate.

There's not much expectation of actually handling `InvalidState` as the caller. Not saying we should, but theoretically, we could attempt to recreate the `UiSurface` for transient issues like a child UI node being promoted to the root UI node.

When I circle back around to this, I'll remove `Unreachable` and revert to `unreachable!`.